alternator/executor.cc

```diff
@@ -258,8 +258,8 @@ schema_ptr executor::find_table(service::storage_proxy& proxy, const rjson::valu
     }
 }
 
-schema_ptr get_table(service::storage_proxy& proxy, const rjson::value& request) {
-    auto schema = executor::find_table(proxy, request);
+schema_ptr executor::get_table(const rjson::value& request) {
```

Comment:
There is beauty in things you mentioned I agree, also there is beauty in the code which guides the reader, a code which can self explain its purpose and structure. It should start simple and then unfold layer upon layer. Alternator does start simple and I think it's a good piece of code, but later it unfolds very quickly into a bigger set of freely floating functions and local micro classes.

I think in this case I failed to preserve the original design as I wasn't able to see it because there was too much stuff I couldn't explain to myself, for instance:
- executor class has no comment about it's purpose
- executor contains `static schema_ptr find_table` which is used by freely floating non static `schema_ptr get_table`, why?
- executor has `friend class rmw_operation` why? Isn't it cyclic dependency already?
- rmw_operation is confusing by having both `static` methods and non static fields. So is this class stateless or not ? (it's not)
-  why do stuff like that `_returnvalues(parse_returnvalues(_request))` ? When reading `rmw_operation::returnvalues rmw_operation::parse_returnvalues(const rjson::value& request)` it requires few mental hops to realize that both input and output of this static function actually belong to `rmw_operation` state (should this function be even protected instead of private?)
- filter is a facade class and underneath it can be `std::optional<std::variant<conditions_filter, expression_filter>> _imp` so either conditions_filter, expression_filter, or no filter, but why implement inheritance using composition with std::variant and std::visit ? Is it for performance reasons?

And more general question if the implementation is put in single executor.cc file and consists of static functions mixed with small standalone classes spread over cc and hh with the main principle that everything is as local as possible (except when this randomly breaks, e.g. `class attribute_path_map_node`), how the reader is supposed to know what is connected with what?

Reply:
You're right - whatever were the original design goals, several years of "quick fixes", hacks and improvements took their toll too. I have issue https://github.com/scylladb/scylladb/issues/5783 about, among other things, splitting Alternator to more meaningful files than just half of it in "executor.cc".

I'll try to answer some specific questios:
Executor was, at least originally, meant to hold the top-level **methods** that invoke the various DynamoDB API requests - PutItem, GetItem, CreateTable, and so on and so on. The source file alternator.cc also holds a lot of implementation details of these operations - all of it file-local ("static"), and #5783 is about splitting this implementation to smaller files (e.g., the write operations share a lot of code but doesn't share a lot of code with the read operations, so we can have a separate executor_write.cc, etc.). In any case, executor.hh - the header file - used to list only the toplevel DynamoDB API requests - those PutItem, GetItem, etc. It didn't used to be a grab-bag of random functions which are useful from random places in the code. But I agree we didn't religiously uphold this, and things like find_table "happened" (you can use "git blame" to see when and how) :-(

rmw_operation has state (it's a superclass of objects that track do read-modify-write operations), and has fair amounts of comments explaining it. It has two static functions which truely do what static functions do in a class (work on a class-global variable - the default write isolation), and one static convenient function to parse the write isolation, which I agree didn't need to be in the class, but putting it in the same header file at least makes sense.

I didn't understand the question about `_returnvalues(parse_returnvalues(_request))`. That's just a C++ way of assigning a local variable in the constructor. We could have assigned it inside the constructor's body, which I agree would have been clearer, but then at least theoretically C++ would assign this variable twice which is 0.0001% slower ;-)

You're right about "class filter". I could have done it with inheritance instead of std::variant. I guess the C++11 (re)training led me to think more in the direction of std::variant instead of inheritance. But it's more than that. We've had a terrible experience (ask @cvybhu) with how Cassandra implemented expressions using inheritence. Althought he function `filter::check()` is easy to implement either way, as soon as you start adding more functions like `filter::filters_on()` and others (see executor.cc), you find yourself touching many classes, often adding implementations that make no sense in some of the classes. In the case of Alternator's "class filter" I admit that both approaches will work well.

Regarding the huge executor.cc containing the implementation of different unrelated things, I agree, and want to split it - as I mentioned this https://github.com/scylladb/scylladb/issues/5783